storm-core/src/jvm/org/apache/storm/assignments/AssignmentDistributionService.java

```diff
@@ -0,0 +1,232 @@
+ * <p>We support multiple working threads to distribute assignment, every thread has a queue buffer.
+ * <p>Master will shuffle its node request to the queues, if the target queue is full, we just discard the request, let the supervisors sync instead.
+ * <p/>
+ * <pre>{@code
+ * Working mode
+ *                      +--------+         +-----------------+
+ *                      | queue1 |   ==>   | Working thread1 |
+ * +--------+ shuffle   +--------+         +-----------------+
+ * | Master |   ==>
+ * +--------+           +--------+         +-----------------+
+ *                      | queue2 |   ==>   | Working thread2 |
+ *                      +--------+         +-----------------+
+ * }
+ * </pre>
+ */
+public class AssignmentDistributionService implements Closeable {
+    private static final Logger LOG = LoggerFactory.getLogger(AssignmentDistributionService.class);
+    private ExecutorService service;
+
+    /**
+     * Flag to indicate if the service is active
+     */
+    private volatile boolean active = false;
+
+    private Random random;
+    /**
+     * Working threads num.
+     */
+    private int threadsNum = 0;
+    /**
+     * Working thread queue size.
+     */
+    private int queueSize = 0;
+
+    /**
+     * Assignments request queue.
+     */
+    private volatile Map<Integer, LinkedBlockingQueue<NodeAssignments>> assignmentsQueue;
+
+    private Map conf;
+
+    /**
+     * Function for initialization.
+     *
+     * @param conf
+     */
+    public void prepare(Map conf) {
+        this.conf = conf;
+        this.random = new Random(47);
+
+        this.threadsNum = Utils.getInt(conf.get(Config.NIMBUS_ASSIGNMENTS_SERVICE_THREADS), 10);
+        this.queueSize = Utils.getInt(conf.get(Config.NIMBUS_ASSIGNMENTS_SERVICE_THREAD_QUEUE_SIZE), 100);
+
+        this.assignmentsQueue = new HashMap<>();
+        for (int i = 0; i < threadsNum; i++) {
+            this.assignmentsQueue.put(i, new LinkedBlockingQueue<NodeAssignments>(queueSize));
+        }
+        //start the thread pool
+        this.service = Executors.newFixedThreadPool(threadsNum);
+        this.active = true;
+        //start the threads
+        for (int i = 0; i < threadsNum; i++) {
+            this.service.submit(new DistributeTask(this, i));
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.active = false;
+        this.service.shutdownNow();
+        try {
+            this.service.awaitTermination(10l, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            LOG.error("Failed to close assignments distribute service");
+        }
+        this.assignmentsQueue = null;
+    }
+
+    public void addAssignmentsForNode(String node, SupervisorAssignments assignments) {
+        try {
+            boolean success = nextQueue().offer(NodeAssignments.getInstance(node, assignments), 5l, TimeUnit.SECONDS);
```

Comment:
So we have a 5 second timeout on offering, instead of a 5 second timeout on connecting and sending?

Reply:
i also config the SupervisorClient timeout to 5s timeout, which is taken in ThriftConnectionType.